src/data/buffer/basic.lean

```diff
@@ -42,18 +42,120 @@ lemma append_list_mk_buffer  :
 by ext x : 1; simp [array.to_buffer,to_list,to_list_append_list];
    induction xs; [refl,skip]; simp [to_array]; refl
 
+@[simp] lemma to_buffer_to_list (b : buffer α) : b.to_list.to_buffer = b :=
+begin
+  cases b,
+  simp [to_list,to_array,list.to_buffer],
+  congr, simp, refl, apply array.to_list_to_array
+end
+
+@[simp] lemma to_list_to_buffer (l : list α) : l.to_buffer.to_list = l :=
+begin
+  induction l with hd tl hl, refl,
+  simp [list.to_buffer,append_list],
+  rw ← hl, refl
+end
+
+@[simp] lemma append_list_nil (b : buffer α) : b.append_list [] = b := rfl
+
+lemma to_buffer_cons (c : α) (l : list α) :
+  (c :: l).to_buffer = [c].to_buffer.append_list l :=
+begin
+  induction l with hd tl hl,
+  { simp },
+  { apply ext,
+    simp [hl] }
+end
+
+@[simp] lemma size_push_back (b : buffer α) (a : α) : (b.push_back a).size = b.size + 1 :=
+by { cases b, simp [size, push_back] }
+
+@[simp] lemma size_append_list (b : buffer α) (l : list α) :
+  (b.append_list l).size = b.size + l.length :=
+begin
+  induction l with hd tl hl generalizing b,
+  { simp },
+  { simp [append_list, hl, add_comm, add_assoc] }
+end
+
+@[simp] lemma size_to_buffer (l : list α) : l.to_buffer.size = l.length :=
+begin
+  induction l with hd tl hl,
+  { simpa },
+  { rw [to_buffer_cons],
+    have : [hd].to_buffer.size = 1 := rfl,
+    simp [add_comm, this] }
+end
+
+lemma size_singleton (a : α) : [a].to_buffer.size = 1 := rfl
+
+@[simp] lemma read_push_back_left (b : buffer α) (a : α) {i : ℕ} (h : i < b.size) :
```

Comment:
I doubt this lemma can be used by `simp`. Is it used by any of your proofs?

Reply:
I've made the usage explicit below.

Reply:
Am I missing something? `simp` will do conditional rewriting just fine -- if `i < b.size` is itself provable by `simp` then this lemma will fire. This can be expensive if abused but since the head symbol here (`buffer.read`) isn't exactly common, I wouldn't worry too much here.

That's just to say this *can* be used by `simp`, not that it *should*. `simp` will only prove very straightforward inequalities, this will mostly only fire when `i < b.size` is in the simp set, which may not be often depending on your use case.